### ISSUE TITLE
kiam and kube2iam inconsistency

### ISSUE THREAD

**Issue by Arttii** [uswitch/kiam#97]
We are getting a very weird error switching kube2iam for kiam. It works for almost all our service, but in one of our clusters we have Artifactory deployed, which seems to be using ```https://bitbucket.org/jmurty/jets3t/src/default/src/org/jets3t/service/security/AWSEC2IAMSessionCredentials.java``` for API access. 
When using kiam the credentials cannot be fetched without any meaningful error, but it works when using kube2iam. If i go inside the pod, I can curl the metadata api and get the temporary credentials in both cases. 

The only visible difference is a single header:

```

kube2iam

HTTP/1.1 200 OK
Server: EC2ws
Date: Thu, 21 Jun 2018 12:35:06 GMT
Content-Length: 615
Content-Type: text/plain; charset=utf-8

{"AccessKeyId":"","Code":"Success","Expiration":"2018-06-21T12:57:26Z","LastUpdated":"2018-06-21T12:27:26Z","SecretAccessKey":"","Token":"","Type":"AWS-HMAC"}

kiam 

HTTP/1.1 200 OK
Date: Thu, 21 Jun 2018 12:37:47 GMT
Content-Length: 579
Content-Type: text/plain; charset=utf-8

{"Code":"Success","Type":"AWS-HMAC","AccessKeyId":"","SecretAccessKey":"","Token":"","Expiration":"2018-06-21T12:52:00Z","LastUpdated":"2018-06-21T12:37:00Z"}
```
Is anybody else running into this?

**Comment by pingles** [comment]
@Arttii the difference being the `Server: EC2ws` value?

What instance type are you using?

**Comment by Arttii** [comment]
Masters are m4.large and the nodes are m4.xlarge.

Why do you think that would matter?
________________________________

**Comment by pingles** [comment]
There was another issue where the metadata api differed in behaviour between instance types. Just wanted to check. We're running m4 and r4 instances and don't have a problem so unlikely to be that.

So you're able to retrieve the credentials if you curl metadata from a pod, but using it from jets3t fails?

**Comment by Arttii** [comment]
Yes exactly it is very confusing.

But kube2iam works, with the only difference being the single header. So I started thinking it's that.

Could you point me to the part of the code where the 'reverse' proxying is done. In kube2iam it seems to fully pass the request through.
________________________________

**Comment by pingles** [comment]
All metadata API proxying is done by the Agent, most of the HTTP server code starts here: https://github.com/uswitch/kiam/blob/master/pkg/aws/metadata/server.go

Having said that, the response for credentials is never proxied to the metadata api by the agent- it's intercepted and credentials are obtained via STS instead.

If you're able to obtain credentials via curl and instead it only stops with the client you're using (jets3t rather than one of the aws sdk's) then its probably something that's missed between the client lib and the response kiam serves. It'd be weird if the server header value did make a difference but maybe there's an assumption in there that prevents it. It'd be useful to know what the error is though- does the client not get credentials, does it hang/time out while waiting etc. I think those would only be from artifactory's log though.

Are you able to write a quick demo Java app that uses the same client lib as you reference above and see if you get the same behaviour?

**Comment by Arttii** [comment]
Sadly were using Artifactory from Jfrog, so no idea what it's actually doing on the inside.

I think it's first looking up the role it can assume, judging from the code. And that's where it fails. The error is essentially a failure to fetch session credentials. I can share the logs, a bit later, but they are not really meaningful. That call should be proxies right?


Von: Paul Ingles
Gesendet: Freitag, 22. Juni, 15:29
Betreff: Re: [uswitch/kiam] kiam and kube2iam inconsistency (#97)
An: uswitch/kiam
Cc: Arttii, Mention


All metadata API proxying is done by the Agent, most of the HTTP server code starts here: https://github.com/uswitch/kiam/blob/master/pkg/aws/metadata/server.go
Having said that, the response for credentials is never proxied to the metadata api by the agent- it's intercepted and credentials are obtained via STS instead.
If you're able to obtain credentials via curl and instead it only stops with the client you're using (jets3t rather than one of the aws sdk's) then its probably something that's missed between the client lib and the response kiam serves. It'd be weird if the server header value did make a difference but maybe there's an assumption in there that prevents it. It'd be useful to know what the error is though- does the client not get credentials, does it hang/time out while waiting etc. I think those would only be from artifactory's log though.
Are you able to write a quick demo Java app that uses the same client lib as you reference above and see if you get the same behaviour?

**Comment by pingles** [comment]
It'd be interesting to see the stack trace to be sure its breaking at that point.

To prove whether it's a problem with jets3t you could try whipping up a quick Java CLI app which makes an API call that uses the lib- see whether it fails in the same way as artifactory- to try and isolate where the failure is.

**Comment by Arttii** [comment]
Hi,

So I found it. Looking closely at the jsons, the actual only difference is the Expiration interval. kube2iam has 30 minutes and the kiam one is 15.
And looking at this line ```https://bitbucket.org/jmurty/jets3t/src/default/src/org/jets3t/service/security/AWSEC2IAMSessionCredentials.java?fileviewer=file-view-default#AWSEC2IAMSessionCredentials.java-158``` , they are expecting stuff more than 15 mins, so its constantly retrying. Setting `session-duration` to 30 works out.
 So the issues is solved by  #93. Thanks for your help
